frontend/src/features/turnstile/useTurnstile.tsx

```diff
@@ -0,0 +1,152 @@
+import { useCallback, useEffect, useRef, useState } from 'react'
+import { useIntervalWhen } from 'rooks'
+
+import { featureFlags } from '~shared/constants'
+
+import { useScript } from '~hooks/useScript'
+
+import { useIsFeatureEnabled } from '~features/feature-flags/queries'
+
+type TurnstileBaseConfig = {
+  sitekey?: string
+  theme?: 'light' | 'dark' | 'auto'
+  size?: 'normal' | 'compact'
+  tabindex?: number
+  language?: string
+  execution?: 'default' | 'execute'
+  appearance?: 'always' | 'execute' | 'interaction-only'
+}
+
+interface UseTurnstileProps extends TurnstileBaseConfig {
+  // id of container to load Turnstile captcha in.
+  containerID?: string
+}
+
+type TurnstileConfig = TurnstileBaseConfig & {
+  callback?: (token: string) => void
+  'expired-callback'?: () => void
+  'error-callback'?: () => void
+}
+
+type TurnstileCallbacks = {
+  ready: (callback: () => void) => void
+  render: (containerID?: string, config?: TurnstileConfig) => string
+  reset: (widgetID: string | undefined) => void
+  execute: (widgetID: string) => void
+  getResponse: (widgetID: string) => string
+}
+
+declare global {
+  interface Window {
+    turnstile?: TurnstileCallbacks
+  }
+}
+
+// todo: check usages
+type TurnstileExecutionCallback = {
+  resolve?: (response: string | null) => void
+  reject?: (reason?: unknown) => void
+}
+
+const DEFAULT_CONTAINER_ID = 'turnstile-container'
+
+export const useTurnstile = ({
+  containerID = DEFAULT_CONTAINER_ID,
+  sitekey,
+  execution = 'execute',
+  theme = 'light',
+  appearance = 'interaction-only',
+}: UseTurnstileProps) => {
+  useScript(
+    'https://challenges.cloudflare.com/turnstile/v0/api.js?render=explicit',
+  )
+  const turnstile: TurnstileCallbacks | undefined = window.turnstile
+  const [hasLoaded, setHasLoaded] = useState(turnstile?.render ? true : false)
+  const [widgetID, setWidgetID] = useState<string>()
+
+  const executionPromise = useRef<TurnstileExecutionCallback>({})
+
+  // Feature flag to control turnstile captcha rollout
+  // defaults to false
+  // Cloudflare Turnstile and Google reCaptcha should be mutually exclusive
+  // todo: remove after full rollout
+  const enableTurnstileFeatureFlag = useIsFeatureEnabled(
+    featureFlags.turnstile,
+    false,
+  )
+  useIntervalWhen(
+    () => {
+      if (turnstile?.render) {
+        setHasLoaded(true)
+      }
+    },
+    500,
+    !!sitekey && !hasLoaded,
+  )
+
+  const handleChange = useCallback((response: string | null) => {
+    if (executionPromise.current.resolve) {
+      executionPromise.current.resolve(response)
+      executionPromise.current = {}
+    }
+  }, [])
+
+  const handleError = useCallback(() => {
+    if (executionPromise.current?.reject) {
```

Comment:
Why is there a conditional chaining operator here while there's isn't one for the resolve in the handleChange check above?

Reply:
According to https://developers.cloudflare.com/turnstile/get-started/client-side-rendering/ , handleError is called when there is an error (e.g. network error or the challenge failed). Added conditional chaining in the event that the error is thrown before turnstile.execute is called (i.e. before the promise is set)

Reply:
👍  got it, thanks for the clarification!